In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


# LOADING DATA

In [4]:
data=pd.read_excel('./archivos/DataFramePrueba.xlsb')

In [6]:
from pyxlsb import open_workbook
from openpyxl import Workbook

# Nombre del archivo de origen
input_file = "./archivos/DataFramePrueba.xlsb"  
# Nombres de las hojas a extraer
hojas_a_extraer = ["DataFramePrueba", "diccionario"]  
# Abrir el archivo .xlsb
with open_workbook(input_file) as wb:
    for hoja in hojas_a_extraer:
        try:
            # Acceder a la hoja específica
            with wb.get_sheet(hoja) as sheet:
                # Crear un nuevo archivo de Excel
                nuevo_workbook = Workbook()
                nueva_hoja = nuevo_workbook.active
                nueva_hoja.title = hoja

                # Copiar los datos
                for row in sheet.rows():
                    nueva_hoja.append([cell.v for cell in row])

                # Guardar como archivo .xlsx
                output_file = f"{hoja}.xlsx"
                nuevo_workbook.save(output_file)
                print(f"La hoja '{hoja}' se ha guardado en el archivo '{output_file}'.")
        except Exception as e:
            print(f"Error al procesar la hoja '{hoja}': {e}")


La hoja 'DataFramePrueba' se ha guardado en el archivo 'DataFramePrueba.xlsx'.
La hoja 'diccionario' se ha guardado en el archivo 'diccionario.xlsx'.


# GETTING DATA

In [ ]:
df=pd.read_excel('./archivos/archivosSeparados/DataFramePrueba.xlsx')
diccionario=pd.read_excel('./archivos/archivosSeparados/diccionario.xlsx')

In [11]:
df

IdentificadorCliente               FechaEvento         UsabilidadCupo  \
0                          1  2022-09-19T13:25:31.867Z  0.1184320077740548479   
1                          2  2023-08-23T11:33:46.417Z  0.0771590000000000000   
2                          3  2022-10-01T14:59:48.920Z  0.1855666666666666667   
3                          4  2022-09-22T21:25:09.187Z                   null   
4                          5  2023-03-19T17:48:52.310Z  0.4752148710773535878   
...                      ...                       ...                    ...   
146934                146935  2023-10-20T16:31:14.983Z  0.2875120000000000000   
146935                146936  2023-10-20T20:58:15.520Z                   null   
146936                146937  2023-10-23T22:57:35.243Z                   null   
146937                146938  2023-10-30T15:09:00.610Z                   null   
146938                146939  2023-10-29T19:35:41.283Z  0.9052790000000000000   

         CategoriaPrincipalCredito  DiasMaximosMoraCreditosGenerados  \
0                              NaN                               0.0   
1                  hogar-y-muebles                               NaN   
2                              NaN                             129.0   
3                              NaN                               0.0   
4                      computacion                               0.0   
...                            ...                               ...   
146934  belleza-y-cuidado-personal                               NaN   
146935           electrodomesticos                               NaN   
146936                         NaN                               NaN   
146937                         NaN                               NaN   
146938       celulares-y-telefonos                               NaN   

        NumeroCreditosGPrevius  NumeroCreditosGCanalFPrevius  \
0                          9.0                           9.0   
1                          5.0                           4.0   
2                         33.0                          33.0   
3                          3.0                           3.0   
4                          4.0                           4.0   
...                        ...                           ...   
146934                     NaN                           NaN   
146935                     NaN                           NaN   
146936                     NaN                           NaN   
146937                     NaN                           NaN   
146938                     NaN                           NaN   

        NumeroCreditosGEstadoActivosPrevius  \
0                                       0.0   
1                                       1.0   
2                                       0.0   
3                                       0.0   
4                                       0.0   
...                                     ...   
146934                                  NaN   
146935                                  NaN   
146936                                  NaN   
146937                                  NaN   
146938                                  NaN   

        NumeroCreditosGEstadoPagadosPrevius  NumeroCreditosGCanalVPrevius  \
0                                       9.0                           0.0   
1                                       4.0                           1.0   
2                                      33.0                           0.0   
3                                       3.0                           0.0   
4                                       4.0                           0.0   
...                                     ...                           ...   
146934                                  NaN                           NaN   
146935                                  NaN                           NaN   
146936                                  NaN                           NaN   
146937                                  NaN                       

In [12]:
diccionario

IdentificadorCliente  \
0                              FechaEvento   
1                           UsabilidadCupo   
2                CategoriaPrincipalCredito   
3         DiasMaximosMoraCreditosGenerados   
4                   NumeroCreditosGPrevius   
5             NumeroCreditosGCanalFPrevius   
6             NumeroCreditosGCanalVPrevius   
7      NumeroCreditosGEstadoActivosPrevius   
8      NumeroCreditosGEstadoPagadosPrevius   
9                   NumeroCreditosLPrevius   
10     NumeroCreditosLEstadoActivosPrevius   
11     NumeroCreditosLEstadoPagadosPrevius   
12                 FechaVinculacionCliente   
13                          FechaPrimerUso   
14  TotalPagosEfectuadosGlobalmentePrevius   
15   TotalPagosEfectuadosLocalmentePrevius   
16                  CodigoAlmacenEntregaTC   
17                CodigoMunicipioEntregaTC   
18                  TipoMunicipioEntregaTC   
19                 CanalMunicipioEntregaTC   
20                  NumeroIntentosFallidos   
21                            CupoAprobado   
22                               UsoAppWeb   
23                         ScoreCrediticio   
24                                  Genero   
25                                    Edad   
26                          PerdidaCartera   
27                                DiasMora   

           Numero unico que identifica a cada cliente  
0   Fecha en la cual se hizo la compra en cuestión...  
1   Division entre el valor de la compra y el cupo...  
2   Categoría mayoritaria de la compra que quiere ...  
3   En la historia que lleva el cliente sacando cr...  
4   Cantidad de creditos que ha hecho el cliente a...  
5   Cantidad de creditos que ha hecho el cliente a...  
6   Cantidad de creditos que ha hecho el cliente a...  
7   Cantidad de creditos que ha hecho el cliente a...  
8   Cantidad de creditos que ha hecho el cliente a...  
9   Cantidad de creditos que ha hecho el cliente a...  
10  Cantidad de creditos que ha hecho el cliente e...  
11  Cantidad de creditos que ha hecho el cliente e...  
12  Fecha en la cual se le otorgó la tarjeta de cr...  
13  Fecha en la cual el cliente usó por primera ve...  
14  Pagos que ha realizado el cliente a la tarjeta...  
15  Pagos que ha realizado el cliente a la tarjeta...  
16  Almacen en el cual se le otorgó la tarjeta de ...  
17  municipio en el cual se le otorgó la tarjeta d...  
18  Tipo municipio en el cual se le otorgó la tarj...  
19  Canal por el cual se le otorgó la tarjeta de c...  
20  Cuantas veces el cliente pidió la Tarjeta de C...  
21                      Cupo de la tarjeta de credito  
22  Cual de los canales digitales usa mas el clien...  
23  Puntaje general que califica a un cliente en t...  
24                                 Genero del cliente  
25                                   Edad del cliente  
26  Es el valor que se quiere predecir. 1:El clien...  
27  Indica los dias de Mora que tiene el cliente, ...

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146939 entries, 0 to 146938
Data columns (total 30 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   IdentificadorCliente                    146939 non-null  int64  
 1   FechaEvento                             146939 non-null  object 
 2   UsabilidadCupo                          146939 non-null  object 
 3   CategoriaPrincipalCredito               114111 non-null  object 
 4   DiasMaximosMoraCreditosGenerados        108968 non-null  float64
 5   NumeroCreditosGPrevius                  115051 non-null  float64
 6   NumeroCreditosGCanalFPrevius            115051 non-null  float64
 7   NumeroCreditosGEstadoActivosPrevius     115051 non-null  float64
 8   NumeroCreditosGEstadoPagadosPrevius     115051 non-null  float64
 9   NumeroCreditosGCanalVPrevius            115051 non-null  float64
 10  NumeroCreditosLPrevius                  1150

In [21]:
nulos_por_columna=df.isnull().sum()
columnas_con_nulos = nulos_por_columna[nulos_por_columna > 0] 
total_columnas_con_nulos = len(columnas_con_nulos)  

print(f"Columnas con valores nulos:\n{columnas_con_nulos}")
print(f"Total de columnas con nulos: {total_columnas_con_nulos}")

Columnas con valores nulos:
CategoriaPrincipalCredito                 32828
DiasMaximosMoraCreditosGenerados          37971
NumeroCreditosGPrevius                    31888
NumeroCreditosGCanalFPrevius              31888
NumeroCreditosGEstadoActivosPrevius       31888
NumeroCreditosGEstadoPagadosPrevius       31888
NumeroCreditosGCanalVPrevius              31888
NumeroCreditosLPrevius                    31888
NumeroCreditosLEstadoActivosPrevius       31888
NumeroCreditosLEstadoPagadosPrevius       31888
FechaVinculacionCliente                    1824
FechaPrimerUso                            35874
FechaUltimoUso                            35874
TotalPagosEfectuadosGlobalmentePrevius    39016
TotalPagosEfectuadosLocalmentePrevius     39016
CodigoMunicipioEntregaTC                   1830
TipoMunicipioEntregaTC                     1830
CanalMunicipioEntregaTC                      16
NumeroIntentosFallidos                      128
CupoAprobado                                660
UsoAppWeb   

In [24]:
df.describe()

IdentificadorCliente  DiasMaximosMoraCreditosGenerados  \
count          146939.00000                     108968.000000   
mean            73470.00000                         77.787626   
std             42417.77994                        224.206547   
min                 1.00000                          0.000000   
25%             36735.50000                          0.000000   
50%             73470.00000                          0.000000   
75%            110204.50000                         48.000000   
max            146939.00000                       3309.000000   

       NumeroCreditosGPrevius  NumeroCreditosGCanalFPrevius  \
count           115051.000000                 115051.000000   
mean                12.063155                     11.622046   
std                 17.146242                     17.003708   
min                  0.000000                      0.000000   
25%                  2.000000                      2.000000   
50%                  6.000000                      5.000000   
75%                 15.000000                     15.000000   
max                349.000000                    349.000000   

       NumeroCreditosGEstadoActivosPrevius  \
count                        115051.000000   
mean                              0.292592   
std                               1.086592   
min                               0.000000   
25%                               0.000000   
50%                               0.000000   
75%                               0.000000   
max                              31.000000   

       NumeroCreditosGEstadoPagadosPrevius  NumeroCreditosGCanalVPrevius  \
count                        115051.000000                 115051.000000   
mean                             11.723644                      0.441109   
std                              17.013161                      1.387426   
min                               0.000000                      0.000000   
25%                               2.000000                      0.000000   
50%                               5.000000                      0.000000   
75%                              15.000000                      0.000000   
max                             349.000000                     58.000000   

       NumeroCreditosLPrevius  NumeroCreditosLEstadoActivosPrevius  \
count           115051.000000                        115051.000000   
mean                 0.359754                             0.027058   
std                  1.136460                             0.215295   
min                  0.000000                             0.000000   
25%                  0.000000                             0.000000   
50%                  0.000000                             0.000000   
75%                  0.000000                             0.000000   
max                123.000000                            13.000000   

       NumeroCreditosLEstadoPagadosPrevius  ...  FechaUltimoUso  \
count                        115051.000000  ...   111065.000000   
mean                              0.247177  ...    44838.689425   
std                               0.808488  ...      317.957633   
min                               0.000000  ...    36512.000000   
25%                               0.000000  ...    44717.000000   
50%                               0.000000  ...    44884.000000   
75%                               0.000000  ...    45041.000000   
max                              29.000000  ...    45230.000000   

       TotalPagosEfectuadosGlobalmentePrevius  \
count                           107923.000000   
mean                                34.402481   
std                                 50.421206   
min                                  0.000000   
25%                                  5.000000   
50%                                 16.000000   
75%                                 42.000000   
max                               1102.000000   

       TotalPagosEfectuadosLocalmentePrevius  CodigoMunicipioEntregaTC  \
co

## Descripción superficial de los datos

<div>

<p>
    <ul>
       <li> 146939 filas
       </li> 
       <li> 30 columnas de varios tipos: object,date,float64,int64
       </li>
       <li> 24 columnas con entradas nulas
       </li>
       <li> TotalPagosEfectuadosLocalmentePrevius y  TotalPagosEfectuadosGlobalmentePrevius son las columnas con mas nulos con 39016
       </li>
       <li> Columna a predecir : PerdidaCartera (Si y no)
       </li>    
    </ul>
</p>

</div>


## Análisis de cada columna

<div>

<p>
En esta sección vamos realizar la identificación y conteo de las entidades de cada columna. Esta parte es muy importante a la hora de determinar patrones, realizar segmentaciones y tomar deciciones a la hora de etiquetar, completar datos faltantes o la eliminación de columnas.
</p>
</div>

In [27]:
df.columns

Index(['IdentificadorCliente', 'FechaEvento', 'UsabilidadCupo',
       'CategoriaPrincipalCredito', 'DiasMaximosMoraCreditosGenerados',
       'NumeroCreditosGPrevius', 'NumeroCreditosGCanalFPrevius',
       'NumeroCreditosGEstadoActivosPrevius',
       'NumeroCreditosGEstadoPagadosPrevius', 'NumeroCreditosGCanalVPrevius',
       'NumeroCreditosLPrevius', 'NumeroCreditosLEstadoActivosPrevius',
       'NumeroCreditosLEstadoPagadosPrevius', 'FechaVinculacionCliente',
       'FechaPrimerUso', 'FechaUltimoUso',
       'TotalPagosEfectuadosGlobalmentePrevius',
       'TotalPagosEfectuadosLocalmentePrevius', 'CodigoAlmacenEntregaTC',
       'CodigoMunicipioEntregaTC', 'TipoMunicipioEntregaTC',
       'CanalMunicipioEntregaTC', 'NumeroIntentosFallidos', 'CupoAprobado',
       'UsoAppWeb', 'ScoreCrediticio', 'Genero', 'Edad', 'DiasMora',
       'PerdidaCartera'],
      dtype='object')

In [38]:
df[ 'FechaEvento'].value_counts()

FechaEvento
2023-10-29T19:35:41.283Z    1
2022-09-19T13:25:31.867Z    1
2023-08-23T11:33:46.417Z    1
2022-10-01T14:59:48.920Z    1
2022-09-22T21:25:09.187Z    1
                           ..
2023-03-21T22:16:01.433Z    1
2022-10-01T22:10:17.433Z    1
2023-08-31T18:23:56.993Z    1
2022-05-11T15:22:39.510Z    1
2022-05-31T09:22:36.477Z    1
Name: count, Length: 146939, dtype: int64

### Datos sin nulos en ninguna columna

<div>

<p>
Podemos observar que si quitamos los datos nulos en todas las columnas, nos queda aproximadamente la mitad de los datos. Tenemos que evaluar antes de usar o no estos datos si esos nulos tienen alguna interpretación o simplemente es auscencia de datos y podemos prescindir de estos. Para esto debemos revisar columna por columna y analizarlas con base a la definición brindada en el diccionario.

</p>
</div>

In [23]:
df_sin_nulos = df.dropna()
df_sin_nulos

IdentificadorCliente               FechaEvento         UsabilidadCupo  \
4                          5  2023-03-19T17:48:52.310Z  0.4752148710773535878   
10                        11  2023-07-12T12:09:24.737Z  0.1248160000000000000   
23                        24  2023-10-25T14:46:57.737Z  0.3108332753020750192   
37                        38  2022-10-11T10:46:39.873Z  0.1377377311729431286   
43                        44  2023-10-23T12:58:15.907Z  0.0116811037988396770   
...                      ...                       ...                    ...   
146892                146893  2023-09-19T23:31:40.373Z  0.3950762553867783245   
146897                146898  2023-10-02T13:55:53.683Z  0.0511580000000000000   
146898                146899  2023-10-23T17:15:06.223Z   10000000000000000000   
146913                146914  2023-10-22T19:20:12.033Z  0.1865260000000000000   
146930                146931  2023-10-31T18:51:31.197Z  0.4026225000000000000   

         CategoriaPrincipalCredito  DiasMaximosMoraCreditosGenerados  \
4                      computacion                               0.0   
10      electronica,-audio-y-video                               0.0   
23           celulares-y-telefonos                               0.0   
37      belleza-y-cuidado-personal                               0.0   
43                 pines-virtuales                              72.0   
...                            ...                               ...   
146892  belleza-y-cuidado-personal                               0.0   
146897         animales-y-mascotas                               0.0   
146898  belleza-y-cuidado-personal                               0.0   
146913       celulares-y-telefonos                               0.0   
146930       celulares-y-telefonos                               0.0   

        NumeroCreditosGPrevius  NumeroCreditosGCanalFPrevius  \
4                          4.0                           4.0   
10                         7.0                           7.0   
23                        34.0                          32.0   
37                        26.0                          25.0   
43                        51.0                          45.0   
...                        ...                           ...   
146892                     5.0                           5.0   
146897                     2.0                           1.0   
146898                     1.0                           1.0   
146913                     1.0                           1.0   
146930                     1.0                           1.0   

        NumeroCreditosGEstadoActivosPrevius  \
4                                       0.0   
10                                      0.0   
23                                      0.0   
37                                      0.0   
43                                      0.0   
...                                     ...   
146892                                  5.0   
146897                                  1.0   
146898                                  0.0   
146913                                  0.0   
146930                                  0.0   

        NumeroCreditosGEstadoPagadosPrevius  NumeroCreditosGCanalVPrevius  \
4                                       4.0                           0.0   
10                                      7.0                           0.0   
23                                     34.0                           2.0   
37                                     26.0                           1.0   
43                                     49.0                           6.0   
...                                     ...                           ...   
146892                                  0.0                           0.0   
146897                                  1.0                           1.0   
146898                                  1.0                           0.0   
146913                                  1.0                       

<div>

<p>
 
 <ul>
 <li>
 </li>
 </ul>

</p>
</div>